exchanges/exchange.go

```diff
@@ -1496,3 +1497,84 @@ func (b *Base) GetFundingRates(ctx context.Context, request *order.FundingRatesR
 func (b *Base) IsPerpetualFutureCurrency(asset.Item, currency.Pair) (bool, error) {
 	return false, common.ErrNotYetImplemented
 }
+
+// GetKlineBuilder returns a helper for the fetching of candle/kline data for
+// a single request within a pre-determined time window.
+func (b *Base) GetKlineBuilder(pair currency.Pair, a asset.Item, required kline.Interval, start, end time.Time) (*kline.Builder, error) {
+	if pair.IsEmpty() {
+		return nil, currency.ErrCurrencyPairEmpty
+	}
+	if !a.IsValid() {
+		return nil, asset.ErrNotSupported
+	}
+
+	// NOTE: This allows for checking that the required kline interval is
+	// supported by the exchange and/or can be constructed from lower time frame
+	// intervals.
+	request, err := b.Features.Enabled.Kline.Intervals.Construct(required)
+	if err != nil {
+		return nil, err
+	}
+
+	// NOTE: This check is here to make sure a client is notified that using
+	// this functionality will result in error if the total candles cannot be
+	// theoretically retrieved.
+	if count := kline.TotalCandlesPerInterval(start, end, request); count >
+		int64(b.Features.Enabled.Kline.ResultLimit) {
+		return nil, fmt.Errorf("candles count: %d, max limit: %d %w",
+			count,
+			b.Features.Enabled.Kline.ResultLimit,
+			kline.ErrRequestExceedsExchangeLimits)
+	}
+
+	err = b.ValidateKline(pair, a, request)
+	if err != nil {
+		return nil, err
+	}
+
+	formatted, err := b.FormatExchangeCurrency(pair, a)
+	if err != nil {
+		return nil, err
+	}
+
+	return kline.GetBuilder(b.Name, pair, formatted, a, required, request, start, end)
```

Comment:
I like this concept. I think it could just be called `CreateKlineRequest(...)` and it returns a `kline.Request` or along those lines. 

The distinction between Extended and not would be a bool param which determines whether a Ranges is built. Then both GetKlineCandles and GetKlineCandlesExtended can both run off a `KlineRequest` pointer which you can run a validate on.

Further to this, you can add in a `bool` such as `AllowCandleScaling` which will allow you to use the `aligned` property when requesting candles

Reply:
I have converted all builder naming convestions to kline.Request. 

I would like to keep the distinction between extended and basic functionality as segregated by type based on reducing implementation complexity. For certain functions in wrapper implementations I was thinking more methods and fields exposed, also most likely nil pointers as well, will cause confusion(?). If you can give me some code examples on implementation using a bool happy to reverse my initial thoughts and we can work through it on getting it changerino'd.

As for the aligned property I think I may have done a whoopsie here and if you can check it out. [HERE](https://github.com/thrasher-corp/gocryptotrader/pull/1091/files#diff-c180a8fa329c99efe5f311a56a1554cdf0bed8fdfef0b664a5fac31ce6f50ed1R58-R60)
I am actually forcing alignment on requests, end time seems incorrect because I am actually cutting off the end date so I will push an update for that. I am thinking this might have to be cut out completely in another PR as we should reject start or end based on if alignment is out(?).

Reply:
Nice! The joining isn't super important and quite likely not worth any of the additional complexity. Technically we could have only one wrapper function for retrieving candles which just changes behavior based on the request, but wouldn't be worth the time invested.

I'll chat to you soon to understand the alignment issue as I don't think I fully understand it yet. I would find an error  if it is not the exact correct times or have a bool that allows for rounding as acceptable solutions. Either way, API responses may still return data outside those ranges if they treat the times as inclusive

Reply:
I have added an updated version of what I was trying to do. Especially when we are automatically constructing from lower candles like 5 min, 1min -> 1 hr candles I was thinking this would be important. [Change](https://github.com/thrasher-corp/gocryptotrader/pull/1091/files#diff-c180a8fa329c99efe5f311a56a1554cdf0bed8fdfef0b664a5fac31ce6f50ed1R69-R88)